src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -158,6 +222,140 @@ async def generate_lockfile(
         ),
     )
 
+
+@rule
+async def generate_updated_lockfile(
+    update_req: UpdatePythonLockfileRequest,
+    generate_lockfiles_subsystem: GenerateLockfilesSubsystem,
+    pex_lock_subsystem: PexLockSubsystem,
+    python_setup: PythonSetup,
+) -> ProcessResult:
+    req = update_req.gen_req
+    resolve_config = await Get(ResolvePexConfig, ResolvePexConfigRequest(req.resolve_name))
+
+    for project in pex_lock_subsystem.project:
+        if PipRequirement.parse(project).project_name != project:
+            raise ValueError(
+                f"project {project} is not a bare name; specifier, markers must be specified in Pants targets"
+            )
+
+    # Assuming that any lockfile we are trying to update has a destination
+    # that is a file and not an embedded read only resource.  See
+    # pex_requirements.py
+    maybe_lockfile = await Get(
+        Digest,
+        PathGlobs(
+            [req.lockfile_dest],
+            glob_match_error_behavior=GlobMatchErrorBehavior.ignore,
+        ),
+    )
+    if maybe_lockfile.serialized_bytes_length == 0:
+        raise ValueError(
+            f"Empty or non-existent lockfile at {req.lockfile_dest}. Unable to update in place"
+        )
+
+    original_loaded_lockfile = await Get(
+        LoadedLockfile,
+        LoadedLockfileRequest(
+            Lockfile(
+                url=req.lockfile_dest,
+                url_description_of_origin=f"existing lockfile from {req.resolve_name}",
+                resolve_name=req.resolve_name,
+            )
+        ),
+    )
+
+    if not (
+        original_loaded_lockfile.metadata
+        and isinstance(original_loaded_lockfile.metadata, (PythonLockfileMetadataV3))
+    ):
+        raise ValueError(
+            f"Pants metadata for lockfile at {req.lockfile_dest} is in old <V3 format. Unable to update in place."
+        )
+    elif set(resolve_config.no_binary) != original_loaded_lockfile.metadata.no_binary:
+        raise ValueError(
+            f"Resolve Config no_binary {set(resolve_config.no_binary)} does not match {original_loaded_lockfile.metadata.no_binary} in last lockfile, can not update in place"
+        )
+    elif set(resolve_config.only_binary) != original_loaded_lockfile.metadata.only_binary:
```

Comment:
For these `no_binary`/`only_binary` options, what happens if one is adding a new library, and it needs to be in one of these sets from the start? (I could imagine this applies to a new transitive dependency too, that's not necessarily in the list of requirements the user provides.)

Reply:
The the way Pants handles per package `binary` with `__pip__args.txt` and relying on pip's(?") handling of arguments in cli arguments in requirements files is something... I admit I'm surprised works. I think the answer today is "yes if you change binary for single package you need to regenerate the whole thing."  I don't know the history of why it was done this way or if per-pkg binary would be  reasonable PEX feature.

Reply:
per-package binary only (and per-package source only) would be a reasonable PEX feature. You can be assured the reason it was done this way was to avoid having to wade that deep; since Pex itself tries to not get between you and Pip as much as possible, `pex -r requirements.txt` Actually corresponds to `pip <subcommand> -r requirements.txt` and Pip suppports passing certain arguments via requirements file.

Reply:
This is available as `--only-wheel` and `--only-build` (#2346) and will be released in Pex 2.1.161 shortly.

Reply:
And [Pex 2.1.161](https://github.com/pantsbuild/pex/releases/tag/v2.1.161) with support for `--only-wheel` / `--only-build` is now available.